File: Problems/2818/alternative_code.py
class Solution(object):
    def maximumScore(self, nums, k):
        MOD = 10**9 + 7
        prime_factor_counts = [self._prime_factor_count(n) for n in nums]
        sorted_nums = sorted([(nums[i], i) for i in range(len(nums))],
                             key=lambda x: x[0],
                             reverse=True)
        score = 1
        operations_so_far = 0
        pivot = 0
        
        while operations_so_far < k and pivot < len(sorted_nums):
            possible_subarrays = min(
                self._possible_subarrays(prime_factor_counts, sorted_nums[pivot][1]),
                k - operations_so_far
            )
            score = (score * pow(sorted_nums[pivot][0], possible_subarrays, MOD)) % MOD
            operations_so_far += possible_subarrays
            pivot += 1
        return score % MOD

    def _prime_factor_count(self, n):
        count = 0
        factor = 2
        while factor * factor <= n:
            if n % factor == 0:
                count += 1
                while n % factor == 0:
                    n //= factor
            factor += 1 if factor == 2 else 2
        if n > 1:
            count += 1
        return count

    def _possible_subarrays(self, prime_factor_counts, index):
        nums_length = len(prime_factor_counts)
        index_prime_val = prime_factor_counts[index]

        left, right = index, index
        while left > 0 and prime_factor_counts[left - 1] < index_prime_val:
            left -= 1
        while right < nums_length - 1 and prime_factor_counts[right + 1] <= index_prime_val:
            right += 1
        return (right - index + 1) * (index - left + 1)

File: Problems/2818/alternative_code.c
#include <stdio.h>
#include <stdlib.h>

#define MOD 1000000007

static long long modPow(long long base, long long exp) {
    long long result = 1;
    while (exp > 0) {
        if (exp & 1) result = (result * base) % MOD;
        base = (base * base) % MOD;
        exp >>= 1;
    }
    return result;
}

static int primeFactorCount(int n) {
    int count = 0, f = 2;
    while (f * f <= n) {
        if (n % f == 0) {
            count++;
            while (n % f == 0) n /= f;
        }
        f = (f == 2 ? 3 : f + 2);
    }
    if (n > 1) count++;
    return count;
}

static long long possibleSubarrays(int *counts, int size, int idx) {
    int val = counts[idx], l = idx, r = idx;
    while (l > 0 && counts[l - 1] < val) l--;
    while (r < size - 1 && counts[r + 1] <= val) r++;
    return (long long)(r - idx + 1) * (idx - l + 1);
}

int maximumScore(int* nums, int numsSize, int k) {
    int *pf = malloc(numsSize * sizeof(int));
    for (int i = 0; i < numsSize; i++) pf[i] = primeFactorCount(nums[i]);

    typedef struct { int val, idx; } Pair;
    Pair *arr = malloc(numsSize * sizeof(Pair));
    for (int i = 0; i < numsSize; i++) {
        arr[i].val = nums[i]; arr[i].idx = i;
    }

    int compare(const void* a, const void* b) {
        return ((Pair*)b)->val - ((Pair*)a)->val;
    }
    qsort(arr, numsSize, sizeof(Pair), compare);

    long long score = 1, ops = 0;
    for (int p = 0; p < numsSize && ops < k; p++) {
        long long canUse = possibleSubarrays(pf, numsSize, arr[p].idx);
        if (canUse > k - ops) canUse = k - ops;
        score = (score * modPow(arr[p].val, canUse)) % MOD;
        ops += canUse;
    }
    free(pf); free(arr);
    return (int)score;
}